### PR TITLE
singularity and docker container images from CI

### PR DESCRIPTION
**Pull request by truatpasteurdotfr**
Take 2 for a ready to use napari container.

Cheers

Tru

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,6 +32,15 @@ python -m pip install "napari[all]"
 
 If you prefer conda over pip, you can replace the last line with: `conda install -c conda-forge napari`
 
+### containers:
+Docker
+```sh
+docker pull ghcr.io/truatpasteurdotfr/napari:main
```
Comment (tlambert03) [inline README.md]: with this PR, does one still need to use your branch?  Or is that something you used in testing that can be changed now?  Not sure I love the default container source being outside of the napari org

File: README.md [hunk 2]
```diff
@@ -32,6 +32,15 @@ python -m pip install "napari[all]"
...
+```
+Singularity
+```sh
+singularity exec -B /run oras://ghcr.io/truatpasteurdotfr/napari:latest napari -h
```
Comment (tlambert03) [inline README.md]: same here

Comment (truatpasteurdotfr) [reply README.md]: Of course not! I don't expect anyone to use my fork!
You  definitley need to replace ghcr.io/truatpasteurdotfr/napari by ghcr.io/napari/napari as you are building on your CI...
This PR was just meant as a proof of concept "which works for me" :D.

Comment (sofroniewn) [reply README.md]: does that explain why our build test fails on this PR?

```python-traceback
Traceback (most recent call last):
  File "/home/runner/work/napari/napari/bundle.py", line 312, in <module>
    print('created', bundle())
  File "/home/runner/work/napari/napari/bundle.py", line 288, in bundle
    subprocess.check_call(cmd)
  File "/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['briefcase', 'build', '--no-docker']' returned non-zero exit status 200.
Error: Process completed with exit code 1.
```

Comment (tlambert03) [reply README.md]: > does that explain why our build test fails on this PR?

that looks unrelated, I don't think `bundle.py` should be executing any code in the README

> You definitley need to replace ghcr.io/truatpasteurdotfr/napari by ghcr.io/napari/napari as you are building on your CI...

then that should be stated in the docs.  I guarantee you most people will copy and paste if you don't instruct them (i.e. we can't make assumptions about people's level of understanding here)

File: .github/workflows/docker-singularity-publish.yml [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+on:
+#  schedule:
+#    - cron: '31 0 * * *'
+  push:
+    branches: [ main ]
+    # Publish semver tags as releases.
+    tags: [ 'v*.*.*' ]
+  pull_request:
+    branches: [ main ]
```
Comment (tlambert03) [inline .github/workflows/docker-singularity-publish.yml]: do we really want this to run on every PR?

File: README.md (hunk 2 above)
Comment (truatpasteurdotfr) [reply README.md]: Well, I can make a newPR with the modified paths if you want.

Comment (tlambert03) [reply README.md]: i don't understand... can we just update this PR?

Comment (truatpasteurdotfr) [reply README.md]: I don't know how to update a PR... I will merge the latest push and resubmit.

Comment (truatpasteurdotfr) [reply README.md]: I have modified my fork to reflect the URL changes.

File: .github/workflows/docker-singularity-publish.yml (hunk 3 above)
Comment (sofroniewn) [reply .github/workflows/docker-singularity-publish.yml]: probably not, no - maybe just releases or the nightly cron?

Comment (truatpasteurdotfr) [reply .github/workflows/docker-singularity-publish.yml]: modified to "on demand"

File: README.md [hunk 4]
```diff
@@ -32,6 +32,15 @@ python -m pip install "napari[all]"
...
 
+### containers:
+Docker
+```sh
+docker pull ghcr.io/napari/napari:main
+```
+Singularity
+```sh
+singularity exec -B /run oras://ghcr.io/napari/napari:latest napari -h
+```
```
Comment (sofroniewn) [inline README.md]:
```suggestion
```

We're trying to keep some of the details out of the main readme now. These steps could instead be added in https://github.com/napari/napari/blob/main/docs/tutorials/fundamentals/installation.md#next-steps immediately above the final section. That can happen in a later PR

File: .github/workflows/docker-singularity-publish.yml [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      - name: Check out code for the container build
+        uses: actions/checkout@v2
+
+# ----
+
+  build2:
+    needs: build1
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/truatpasteurdotfr/docker-c7-singularity-builder:main
```
Comment (sofroniewn) [inline .github/workflows/docker-singularity-publish.yml:95]: @tlambert03 any concerns here?

**Review by sofroniewn — COMMENTED** [review]
I'm so sorry this has sat so long @truatpasteurdotfr - personally I have no real experience with Singularity so it makes reviewing hard for me. I flagged one final change to the readme. Maybe also @jaimergp can give this a quick review as he may have more experience here and I'd really like to get the green light to merge from someone on the team who understands the code. Thanks for your patience here!!

**Comment by sofroniewn** [comment]
cc also @potating-potato who likely has more experience here too. thanks!!

File: .github/workflows/docker-singularity-publish.yml (hunk 5 above)
Comment (tlambert03) [reply .github/workflows/docker-singularity-publish.yml:95]: Nope this can go in. I agree with you about the readme though

**Comment by liu-ziyang** [comment]
> cc also @potating-potato who likely has more experience here too. thanks!!

Thanks for the cc, but I am afraid I have as little knowledge of singularity containers. In principle this seems fine, worst case scenario if this does not work as well, it doesn't break much because we don't release docker containers regularly anyway.

As long as the windows tests passed I am ok with this merge. (bonus point if we can migrate the base container somehow to napari organization)

**Comment by sofroniewn** [comment]
@truatpasteurdotfr if you can allow edits from maintainers then i can finish this PR up and get it merged, thanks!

**Comment by truatpasteurdotfr** [comment]
maintainer edit: sure, I just don't know how to enale that!

**Comment by sofroniewn** [comment]
> maintainer edit: sure, I just don't know how to enale that!

Here is some guidance 
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

Also if you just remove the changes from the readme then I will merge this - thanks!

**Comment by truatpasteurdotfr** [comment]
[X] checked, but the CI is failing on the docs part?

**Review by sofroniewn — APPROVED** [review]
I think this is now ready. Thanks all!!

**Pull request merged**